pkg/controller/replicaset/replica_set.go

```diff
@@ -794,14 +794,19 @@ func (rsc *ReplicaSetController) syncReplicaSet(ctx context.Context, key string)
 	if manageReplicasErr != nil {
 		return manageReplicasErr
 	}
-	// Resync the ReplicaSet after MinReadySeconds as a last line of defense to guard against clock-skew.
+	// Plan the next availability check as a last line of defense against queue preemption and guard against clock-skew.
 	if updatedRS.Spec.MinReadySeconds > 0 &&
 		updatedRS.Status.ReadyReplicas == *(updatedRS.Spec.Replicas) &&
 		updatedRS.Status.AvailableReplicas != *(updatedRS.Spec.Replicas) {
-		nextSyncInSeconds = ptr.To(int(updatedRS.Spec.MinReadySeconds))
+		if nextCheck := controller.FindMinNextPodAvailabilityCheck(activePods, updatedRS.Spec.MinReadySeconds, rsc.clock.Now()); nextCheck != nil {
```

Comment:
consider using a happy path to avoid the else condition altogether
```go
// Safeguard fallback to MinReadySeconds. It might be used in case we are waiting for some 
// Pods to be created or scheduled while all other Pods are already available.
nextSyncDuration = ptr.To(time.Duration(updatedRS.Spec.MinReadySeconds) * time.Second)

// Compute the next sync based on Pod availability
if nextCheck := controller.FindMinNextPodAvailabilityCheck(activePods, updatedRS.Spec.MinReadySeconds, rsc.clock.Now()); nextCheck != nil {
    nextSyncDuration = ptr.To(*nextCheck )
}
```

Reply:
Thanks, I have simplified the else branch as suggested. Nevertheless, I have also updated the logic to cover more edge cases, so sadly it is a bit more complicated now.